client/beefy/src/worker.rs

```diff
@@ -490,18 +493,18 @@ where
 					// worker won't vote until it witnesses a session change.
 					// Once we'll implement 'initial sync' (catch-up), the worker will be able to
 					// start voting right away.
-					self.handle_finality_notification(notif);
-					future::ready(false)
+					self.handle_finality_notification(&notif);
+					break
 				} else {
 					trace!(target: "beefy", "🥩 Finality notification: {:?}", notif);
 					debug!(target: "beefy", "🥩 Waiting for BEEFY pallet to become available...");
-					future::ready(true)
 				}
-			})
-			.for_each(|_| future::ready(()))
-			.await;
-		// get a new stream that provides _new_ notifications (from here on out)
-		self.finality_notifications = self.client.finality_notification_stream();
+				},
+				_ = gossip_engine => {
+					break
+				}
```

Comment:
This is here to "pump/discard" the `gossip_engine` events, right?

It shouldn't `break` the "wait for pallet loop"...
```suggestion
				_ = gossip_engine => {
					continue
				}
```

Reply:
Yes, this is here to let the gossip engine process its events, but this future can only be ready when the gossip engine's network stream is closed, so I don't think there is really any point in continuing to wait since the gadget will exit *anyway* if you look at the code below:

```rust
    _ = gossip_engine => {
        error!(target: "beefy", "🥩 Gossip engine has terminated.");
        return;
    }
```

Maybe a comment with an explanation or explicitly returning `false` and doing an early return in the caller would be more appropriate here?